pycircleci/api.py

```diff
@@ -933,18 +968,56 @@ def _request(self, verb, endpoint, data=None, api_version=None):
         request_url = "{0}/{1}/{2}".format(self.url, api_version, endpoint)
 
         if verb == GET:
-            resp = self._session.get(request_url, auth=auth, headers=headers)
+            resp = self._session.get(request_url, params=params, auth=auth, headers=headers)
         elif verb == POST:
-            resp = self._session.post(request_url, auth=auth, headers=headers, json=data)
+            resp = self._session.post(request_url, params=params, auth=auth, headers=headers, json=data)
         elif verb == PUT:
-            resp = self._session.put(request_url, auth=auth, headers=headers, json=data)
+            resp = self._session.put(request_url, params=params, auth=auth, headers=headers, json=data)
         elif verb == DELETE:
-            resp = self._session.delete(request_url, auth=auth, headers=headers)
+            resp = self._session.delete(request_url, params=params, auth=auth, headers=headers)
         else:
             raise CircleciError("Invalid verb: {}. Valid values are: {}".format(verb, verbs))
 
         resp.raise_for_status()
         return resp.json()
+    
+    def _request_get_with_depagination(self, endpoint, paginate=False, limit=None, params=None, api_version=API_VER_V2):
+        """Send one or more 'GET' HTTP Request and depaginates the results, up to a limit.
+
+        :param endpoint: API endpoint to 'GET'.
+        :param paginate: If True, repeatedly requests more items from the endpoint until the limit has been reached (or until all results have been fetched). Defaults to False
+        :param limit: Maximum number of items to return in this response. By default returns all the results from many calls to the endpoint, or all the results from a single call to the endpoint, depending on the value for 'paginate'.
+        :param params: Optional query parameters.
+
+        :type params: dict
+
+        :raises requests.exceptions.HTTPError: When response code is not successful.
+
+        :returns: A list of items which are the combined results of the requests made.
+        """
+        def get_items_v2_api(endpoint, paginate, soft_limit, params):
+            returned_results = 0
+
+            while True:
+                resp = self._request("GET", endpoint, params=params, api_version=api_version)
+                yield from resp['items']
+                returned_results += len(resp['items'])
+
+                if not paginate or not resp["next_page_token"] or (soft_limit is not None and returned_results >= soft_limit):
+                    return
+                
+                params["page-token"] = resp["next_page_token"]
+        
+        params = {} if params is None else params.copy()
+
+        if api_version == API_VER_V2:
+            results = list(get_items_v2_api(endpoint, paginate, limit, params))
+        else:
+            raise NotImplementedError("Unable to depaginate API version {} results".format(api_version))
+
+        return results[:limit]
```

Comment:
need a check to guard against `limit is None`

Reply:
if `limit` is `None` then this line returns the whole `results`

Reply:
oh, you're right. did not know about it :))